src/mcp/server/lowlevel/server.py

```diff
@@ -677,7 +715,11 @@ async def _handle_notification(self, notify: Any):
             logger.debug("Dispatching notification of type %s", type(notify).__name__)
 
             try:
-                await handler(notify)
+                sig = inspect.signature(handler)
+                if "session" in sig.parameters:
+                    await handler(notify, current_session_ctx.get())
+                else:
+                    await handler(notify)
```

Comment:
I don't think we should do introspection based switches in our `lowlevel` server code - we should be as explicit as possible. If we're going to have different behavior based on the type of request, we should explicitly have match/case statements that handle those different cases.

Can we instead consistently pass the session into the `handler` like we do for `requests` if we need it? Some handlers just won't use it and that's fine.

For backwards compatibility we could have the `progress_notification` handler have an optional session argument.